tools/validators/instance_validator/tests/telemetry_validator_test.py

```diff
@@ -0,0 +1,176 @@
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an AS IS BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests tools.validators.instance_validator.telemetry_validator"""
+
+from __future__ import absolute_import
+from __future__ import print_function
+
+import os
+import json
+# import threading
+
+from validate import instance_parser
+from validate import telemetry_error
+from validate import telemetry_validator
+from absl.testing import absltest
+
+_TELEMETRY_PATH = os.path.join('.', 'tests', 'fake_telemetry')
+_INSTANCES_PATH = os.path.join('.', 'tests', 'fake_instances', 'GOOD')
+
+_MESSAGE_ATTRIBUTES_PATH_1 = os.path.join(_TELEMETRY_PATH,
+                                          'message_attributes_CHWS_WDT-17.json')
+_MESSAGE_ATTRIBUTES_1 = json.load(open(_MESSAGE_ATTRIBUTES_PATH_1))
+
+_MESSAGE_ATTRIBUTES_PATH_2 = os.path.join(_TELEMETRY_PATH,
+                                          'message_attributes_DMP_EDM-17.json')
+_MESSAGE_ATTRIBUTES_2 = json.load(open(_MESSAGE_ATTRIBUTES_PATH_2))
+
+class FakeMessage(object):
+  def __init__(self, attributes, data):
+    super().__init__()
+    self.attributes = attributes
+    self.data = data
+
+  def ack(self):
+    return NotImplemented
+
+with open(os.path.join(_TELEMETRY_PATH,
+                       'telemetry_good.json')) as file:
+  _MESSAGE_GOOD = FakeMessage(_MESSAGE_ATTRIBUTES_1, file.read())
+
+with open(os.path.join(_TELEMETRY_PATH,
+                       'telemetry_missing_point.json')) as file:
+  _MESSAGE_MISSING_POINT = FakeMessage(_MESSAGE_ATTRIBUTES_1, file.read())
+
+with open(os.path.join(_TELEMETRY_PATH,
+                       'telemetry_missing_present_value.json')) as file:
+  _MESSAGE_MISSING_PRESENT_VALUE = FakeMessage(
+    _MESSAGE_ATTRIBUTES_1, file.read())
+
+with open(os.path.join(_TELEMETRY_PATH,
+                       'telemetry_invalid_state.json')) as file:
+  _MESSAGE_INVALID_STATE = FakeMessage(_MESSAGE_ATTRIBUTES_2, file.read())
+
+with open(os.path.join(_TELEMETRY_PATH,
+                       'telemetry_invalid_number.json')) as file:
+  _MESSAGE_INVALID_NUMBER = FakeMessage(_MESSAGE_ATTRIBUTES_1, file.read())
+
+with open(os.path.join(_TELEMETRY_PATH,
+                       'telemetry_invalid_number_boolean.json')) as file:
+  _MESSAGE_INVALID_NUMBER_BOOLEAN = FakeMessage(
+    _MESSAGE_ATTRIBUTES_1, file.read())
+
+with open(os.path.join(_TELEMETRY_PATH,
+                       'telemetry_multiple_errors.json')) as file:
+  _MESSAGE_MULTIPLE_ERRORS = FakeMessage(_MESSAGE_ATTRIBUTES_1, file.read())
+
+# TODO: fix inconsistency between telemetry parser expecting a string,
```

Comment:
It's fixed in the sense that the tests work, but there's still an inconsistency between classes in the hierarchy that do similar parsing work (Telemetry, InstanceParser) but accept either a file handle or a string read from a file. This gave me some trouble when writing tests since the error messages were cryptic, so I want to look at refactoring those interfaces.